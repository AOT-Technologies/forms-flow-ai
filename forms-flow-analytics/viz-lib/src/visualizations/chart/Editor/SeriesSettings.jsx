import { includes, map, extend, fromPairs } from "lodash";
import React, { useMemo, useCallback } from "react";
import { useDebouncedCallback } from "use-debounce";
import Table from "antd/lib/table";
import Input from "antd/lib/input";
import Radio from "antd/lib/radio";
import { sortableElement } from "react-sortable-hoc";
import { SortableContainer, DragHandle } from "@/components/sortable";
import { EditorPropTypes } from "@/visualizations/prop-types";
import ChartTypeSelect from "./ChartTypeSelect";
import getChartData from "../getChartData";

const SortableBodyRow = sortableElement(props => <tr {...props} />);

function getTableColumns(options, updateSeriesOption, debouncedUpdateSeriesOption) {
  const result = [
    {
      title: "Order",
      dataIndex: "zIndex",
      render: (unused, item) => (
        <span className="series-settings-order">
          <DragHandle />
          {item.zIndex + 1}
        </span>
      ),
    },
    {
      title: "Label",
      dataIndex: "name",
      render: (unused, item) => (
        <Input
          data-test={`Chart.Series.${item.key}.Label`}
          placeholder={item.key}
          defaultValue={item.name}
          onChange={event => debouncedUpdateSeriesOption(item.key, "name", event.target.value)}
        />
      ),
    },
  ];

  if (!includes(["pie", "heatmap"], options.globalSeriesType)) {
    result.push({
      title: "Y Axis",
      dataIndex: "yAxis",
      render: (unused, item) => (
        <Radio.Group
          className="series-settings-y-axis"
          value={item.yAxis === 1 ? 1 : 0}
          onChange={event => updateSeriesOption(item.key, "yAxis", event.target.value)}>
          <Radio value={0} data-test={`Chart.Series.${item.key}.UseLeftAxis`}>
            left
          </Radio>
          <Radio value={1} data-test={`Chart.Series.${item.key}.UseRightAxis`}>
            right
          </Radio>
        </Radio.Group>
      ),
    });
    result.push({
      title: "Type",
      dataIndex: "type",
      render: (unused, item) => (
        <ChartTypeSelect
          data-test={`Chart.Series.${item.key}.Type`}
          dropdownMatchSelectWidth={false}
          value={item.type}
          onChange={value => updateSeriesOption(item.key, "type", value)}
        />
      ),
    });
  }

  return result;
}

export default function SeriesSettings({ options, data, onOptionsChange }) {
  const series = useMemo(
    () =>
      map(
        getChartData(data.rows, options), // returns sorted series
        ({ name }, zIndex) =>
          extend({ key: name, type: options.globalSeriesType }, options.seriesOptions[name], { zIndex })
      ),
    [options, data]
  );

  const handleSortEnd = useCallback(
    ({ oldIndex, newIndex }) => {
      const seriesOptions = [...series];
      seriesOptions.splice(newIndex, 0, ...seriesOptions.splice(oldIndex, 1));
      onOptionsChange({ seriesOptions: fromPairs(map(seriesOptions, ({ key }, zIndex) => [key, { zIndex }])) });
    },
    [onOptionsChange, series]
  );

  const updateSeriesOption = useCallback(
    (key, prop, value) => {
      onOptionsChange({
        seriesOptions: {
          [key]: {
            [prop]: value,
          },
        },
      });
    },
    [onOptionsChange]
  );
  const [debouncedUpdateSeriesOption] = useDebouncedCallback(updateSeriesOption, 200);

  const columns = useMemo(() => getTableColumns(options, updateSeriesOption, debouncedUpdateSeriesOption), [
    options,
    updateSeriesOption,
    debouncedUpdateSeriesOption,
  ]);

  return (
    <SortableContainer
      axis="y"
      lockAxis="y"
      lockToContainerEdges
      useDragHandle
      helperClass="chart-editor-series-dragged-item"
      helperContainer={container => container.querySelector("tbody")}
      onSortEnd={handleSortEnd}
      containerProps={{
        className: "chart-editor-series",
      }}>
      <Table
        dataSource={series}
        columns={columns}
        components={{
          body: {
            row: SortableBodyRow,
          },
        }}
        onRow={item => ({ index: item.zIndex })}
        pagination={false}
      />
    </SortableContainer>
  );
}

SeriesSettings.propTypes = EditorPropTypes;
