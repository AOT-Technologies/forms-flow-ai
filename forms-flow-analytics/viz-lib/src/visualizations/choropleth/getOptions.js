import { merge } from "lodash";
import ColorPalette from "./ColorPalette";

const DEFAULT_OPTIONS = {
  mapType: "countries",
  countryCodeColumn: "",
  countryCodeType: "iso_a3",
  valueColumn: "",
  clusteringMode: "e",
  steps: 5,
  valueFormat: "0,0.00",
  noValuePlaceholder: "N/A",
  colors: {
    min: ColorPalette["Light Blue"],
    max: ColorPalette["Dark Blue"],
    background: ColorPalette.White,
    borders: ColorPalette.White,
    noValue: ColorPalette["Light Gray"],
  },
  legend: {
    visible: true,
    position: "bottom-left",
    alignText: "right",
  },
  tooltip: {
    enabled: true,
    template: "<b>{{ @@name }}</b>: {{ @@value }}",
  },
  popup: {
    enabled: true,
    template: "Country: <b>{{ @@name_long }} ({{ @@iso_a2 }})</b>\n<br>\nValue: <b>{{ @@value }}</b>",
  },
};

export default function getOptions(options) {
  return merge({}, DEFAULT_OPTIONS, options);
}
