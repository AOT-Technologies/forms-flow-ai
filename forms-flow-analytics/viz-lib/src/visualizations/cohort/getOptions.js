import { merge } from "lodash";
import ColorPalette from "@/visualizations/ColorPalette";

const DEFAULT_OPTIONS = {
  timeInterval: "daily",
  mode: "diagonal",
  dateColumn: "date",
  stageColumn: "day_number",
  totalColumn: "total",
  valueColumn: "value",

  showTooltips: true,
  percentValues: true,

  timeColumnTitle: "Time",
  peopleColumnTitle: "Users",
  stageColumnTitle: "{{ @ }}",

  numberFormat: "0,0[.]00",
  percentFormat: "0.00%",
  noValuePlaceholder: "-",

  colors: {
    min: "#ffffff",
    max: ColorPalette["Dark Blue"],
    steps: 7,
  },
};

export default function getOptions(options) {
  return merge({}, DEFAULT_OPTIONS, options);
}
