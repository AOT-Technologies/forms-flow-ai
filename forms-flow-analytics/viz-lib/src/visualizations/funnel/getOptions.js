import { isFinite, map, merge, includes } from "lodash";

const DEFAULT_OPTIONS = {
  stepCol: { colName: null, displayAs: "Steps" },
  valueCol: { colName: null, displayAs: "Value" },
  autoSort: true,
  sortKeyCol: { colName: null, reverse: false },
  itemsLimit: 100,
  percentValuesRange: { min: 0.01, max: 1000.0 },
  numberFormat: "0,0[.]00",
  percentFormat: "0[.]00%",
};

export default function getOptions(options, { columns }) {
  options = merge({}, DEFAULT_OPTIONS, options);

  // Validate
  const availableColumns = map(columns, c => c.name);
  if (!includes(availableColumns, options.stepCol.colName)) {
    options.stepCol.colName = null;
  }
  if (!includes(availableColumns, options.valueCol.colName)) {
    options.valueCol.colName = null;
  }
  if (!includes(availableColumns, options.sortKeyCol.colName)) {
    options.sortKeyCol.colName = null;
  }

  if (!isFinite(options.itemsLimit)) {
    options.itemsLimit = DEFAULT_OPTIONS.itemsLimit;
  }
  if (options.itemsLimit < 2) {
    options.itemsLimit = 2;
  }

  if (options.autoSort) {
    options.sortKeyCol.colName = options.valueCol.colName;
    options.sortKeyCol.reverse = true;
  }

  return options;
}
