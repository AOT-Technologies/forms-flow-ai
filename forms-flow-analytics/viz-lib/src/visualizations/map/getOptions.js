import { merge } from "lodash";

const DEFAULT_OPTIONS = {
  latColName: "lat",
  lonColName: "lon",
  classify: null,
  groups: {},
  mapTileUrl: "//{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
  clusterMarkers: true,
  customizeMarkers: false,
  iconShape: "marker",
  iconFont: "circle",
  foregroundColor: "#ffffff",
  backgroundColor: "#356AFF",
  borderColor: "#356AFF",
  bounds: null,
  tooltip: {
    enabled: false,
    template: "",
  },
  popup: {
    enabled: true,
    template: "",
  },
};

export default function getOptions(options) {
  options = merge({}, DEFAULT_OPTIONS, options);
  options.mapTileUrl = options.mapTileUrl || DEFAULT_OPTIONS.mapTileUrl;

  // Backward compatibility
  if (options.classify === "none") {
    options.classify = null;
  }

  return options;
}
