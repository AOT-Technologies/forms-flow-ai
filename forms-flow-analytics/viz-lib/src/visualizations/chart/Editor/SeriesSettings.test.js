import React from "react";
import enzyme from "enzyme";

import getOptions from "../getOptions";
import SeriesSettings from "./SeriesSettings";

function findByTestID(wrapper, testId) {
  return wrapper.find(`[data-test="${testId}"]`);
}

function mount(options, done) {
  options = getOptions(options);
  return enzyme.mount(
    <SeriesSettings
      visualizationName="Test"
      data={{ columns: [{ name: "a", type: "string" }], rows: [{ a: "test" }] }}
      options={options}
      onOptionsChange={changedOptions => {
        expect(changedOptions).toMatchSnapshot();
        done();
      }}
    />
  );
}

describe("Visualizations -> Chart -> Editor -> Series Settings", () => {
  test("Changes series type", done => {
    const el = mount(
      {
        globalSeriesType: "column",
        columnMapping: { a: "y" },
        seriesOptions: {
          a: { type: "column", label: "a", yAxis: 0 },
        },
      },
      done
    );

    findByTestID(el, "Chart.Series.a.Type")
      .last()
      .simulate("click");
    findByTestID(el, "Chart.ChartType.area")
      .last()
      .simulate("click");
  });

  test("Changes series label", done => {
    const el = mount(
      {
        globalSeriesType: "column",
        columnMapping: { a: "y" },
        seriesOptions: {
          a: { type: "column", label: "a", yAxis: 0 },
        },
      },
      done
    );

    findByTestID(el, "Chart.Series.a.Label")
      .last()
      .simulate("change", { target: { value: "test" } });
  });

  test("Changes series axis", done => {
    const el = mount(
      {
        globalSeriesType: "column",
        columnMapping: { a: "y" },
        seriesOptions: {
          a: { type: "column", name: "a", yAxis: 0 },
        },
      },
      done
    );

    findByTestID(el, "Chart.Series.a.UseRightAxis")
      .last()
      .find("input")
      .simulate("change", { target: { checked: true } });
  });
});
