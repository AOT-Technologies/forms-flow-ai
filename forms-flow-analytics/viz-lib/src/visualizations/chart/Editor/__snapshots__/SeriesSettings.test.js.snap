// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visualizations -> Chart -> Editor -> Series Settings Changes series axis 1`] = `
Object {
  "seriesOptions": Object {
    "a": Object {
      "yAxis": 1,
    },
  },
}
`;

exports[`Visualizations -> Chart -> Editor -> Series Settings Changes series label 1`] = `
Object {
  "seriesOptions": Object {
    "a": Object {
      "name": "test",
    },
  },
}
`;

exports[`Visualizations -> Chart -> Editor -> Series Settings Changes series type 1`] = `
Object {
  "seriesOptions": Object {
    "a": Object {
      "type": "area",
    },
  },
}
`;
